lib/authorizer/ntlm.js

```diff
@@ -0,0 +1,136 @@
+    },
+
+    /**
+     * Verifies whether the request has correct NTLM credentials.
+     *
+     * @param {RequestContext} context
+     * @param {Requester} requester
+     * @param done
+     */
+    pre: function (context, requester, done) {
+        if (!this._state) {
+            return done(null, false);
+        }
+        done(null, true);
+    },
+
+    /**
+     * Verifies whether the no auth succeeded. (Which means, does nothing.)
+     *
+     * @param {RequestContext} context
+     * @param {Requester} requester
+     * @param done
+     */
+    post: function (context, requester, done) {
+        var auth = this,
+            typeOneMessage,
+            typeTwoMessage,
+            typeThreeMessage,
+            response = context.response,
+            item = context.item;
+
+        if (response.code !== 401 && response.code !== 403) {
+            return done(null, true);
+        }
+
+        if (auth._state === STATES.INITIALIZED) {
+
+            // Nothing to do if the server does not ask us for auth in the first place.
+            if (!(response.headers.has('www-authenticate', 'NTLM') &&
+                  response.headers.has('www-authenticate', 'Negotiate'))) {
+                return done(null, true);
+            }
+
+            // Create a type 1 message to send to the server
+            typeOneMessage = ntlmUtil.createType1Message({
+                domain: auth.domain,
+                workstation: auth.workstation || ''
+            });
+
+            // Add the type 1 message as the auth header
+            item.request.upsertHeader({
+                key: 'Authorization',
+                value: typeOneMessage
+            });
+
+            // Update the state
+            auth._state = STATES.T1_MSG_CREATED;
+
+            // ask runtime to replay the request
+            return done(null, false);
+        }
+        else if (auth._state === STATES.T1_MSG_CREATED) {
+
+            // At this point, we can assume that the type 1 message was sent to the server
+            typeTwoMessage = ntlmUtil.parseType2Message(response.headers.get('www-authenticate') || '');
+
+            if (!typeTwoMessage) {
+                return done(new Error('ntlm: server did not correctly process authentication request'));
+            }
+
+            typeThreeMessage = ntlmUtil.createType3Message(typeTwoMessage, {
+                domain: auth.domain,
+                workstation: auth.workstation || '',
+                username: auth.username,
+                password: auth.password
+            });
+
+            // Now create the type 3 message, and add it to the request
+            item.request.upsertHeader({
+                key: 'Authorization',
+                value: typeThreeMessage
+            });
+
+            auth._state = STATES.T3_MSG_CREATED;
+
+            // ask runtime to replay the request
+            return done(null, false);
+        }
+        else if (auth._state === STATES.T3_MSG_CREATED) {
```

Comment:
This case is the same as the default, and should be dropped (unless added to demonstrate intent)